File: 0187_repeated_dna_sequences/rolling_hash/go/solution.go
package repeateddnasequences

const (
	windowSize = 10
	windowMask = (1 << (windowSize * 2)) - 1
)

func encode(ch byte) int {
	switch ch {
	case 'A':
		return 0
	case 'C':
		return 1
	case 'G':
		return 2
	default:
		return 3
	}
}

func findRepeatedDnaSequences(s string) []string {
	if len(s) < windowSize {
		return []string{}
	}

	rolling := 0
	seen := make(map[int]struct{}, len(s))
	repeated := make(map[int]struct{})
	result := make([]string, 0)

	for index := range s {
		rolling = ((rolling << 2) | encode(s[index])) & windowMask
		if index < windowSize-1 {
			continue
		}

		start := index - windowSize + 1
		if _, ok := seen[rolling]; ok {
			if _, alreadyAdded := repeated[rolling]; !alreadyAdded {
				repeated[rolling] = struct{}{}
				result = append(result, s[start:index+1])
			}
			continue
		}
		seen[rolling] = struct{}{}
	}

	return result
}

File: 0187_repeated_dna_sequences/rolling_hash/python/solution.py
WINDOW_SIZE = 10
WINDOW_MASK = (1 << (WINDOW_SIZE * 2)) - 1
ENCODE = {"A": 0, "C": 1, "G": 2, "T": 3}


def find_repeated_dna_sequences(s: str) -> list[str]:
    if len(s) < WINDOW_SIZE:
        return []

    rolling = 0
    seen: set[int] = set()
    repeated: set[int] = set()
    result: list[str] = []

    for index, ch in enumerate(s):
        rolling = ((rolling << 2) | ENCODE[ch]) & WINDOW_MASK
        if index < WINDOW_SIZE - 1:
            continue

        start = index - WINDOW_SIZE + 1
        if rolling in seen:
            if rolling not in repeated:
                repeated.add(rolling)
                result.append(s[start:index + 1])
        else:
            seen.add(rolling)

    return result
